Add JFR old object sample event

galderz opened the pull request:
Implements JFR `jdk.OldObjectSample` event for #5145, without path to GC root functionality. I have a prototype for the path to GC root functionality, but I’ve decided to split this work into a separate PR to make reviewing more manageable.

This PR includes:
* Event emitted for sampled objects and arrays that are possible leaks.
* They are sampled when they’re allocated via the slow path. The slow path happens when an object does not fit in the TLAB and a new one is required. Arrays bigger than the large array threshold are also allocated via the slow path regardless.
* Following the same logic found in HotSpot, the sampler is implemented as a custom priority queue ordered by span, which is based on allocation size. Objects with the smallest span are at the top of the queue.
* The sample queue is initialized with a fixed-size array at runtime. The default size is 256 elements, but this can be configured at runtime using `-XX:FlightRecorderOptions=old-object-queue-size=<size>` option.
* If the queue is full, and the sampled object is bigger than the top, the top is removed and the sampled object is inserted.
* If the queue is full and the sampled object is smaller than the top, the code tries to scavenge the queue which means purging objects that have been garbage collected.
* To scavenge, the underlying array backing the queue is iterated over, removing those sampled objects that have been garbage collected.
* To check if an object has been garbage collected or not, sampled objects are wrapped around WeakReferences.
* Operations on the sampler queue are protected by a new non-reentrant `SpinLock`. This new lock implementation works similarly to a `VMMutex` but extends it to offer a `tryLock()` function. This function is useful for sampling because if another object is already being sampled and the lock cannot be acquired, the sampling is simply skipped. When emitting old object sample events, the sampler waits to acquire the lock.
* Old object sample events emitted with custom descriptions for `Thread`, `ThreadGroup` and `Class`. Ellipsis descriptions are also supported.

Caveats:
* ~Calling `WeakReference.get()` from uninterruptible code fails. Adding `@AnnotateOriginal @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)` to `Target_java_lang_ref_Reference.get()` does not seem to have any impact. Looking at `AnnotationSubstitutionProcessor`, classes that have `@Substitute` don’t seem to get processed for methods with `@AnnotateOriginal`. Worked around by adding `calleeMustBe = false` to the `Uninterruptible` method within which `WeakReference.get()` is called.~ This caveat has been addressed, see latest commits.
* Old object queue size can only be configured at runtime via the startup recorder options flag. I couldn’t find a way to unit test this since all unit tests run as a single binary run, so the queue size is fixed for all the tests.
* Old object sample event descriptions for classes that have a `size` field not implemented since they’re not feasible in a closed world environment. In HotSpot this is achieved by using reflection to lookup a `size` field but that cannot be done in SubstrateVM.
* Old object sample event description tests only added for thread group. Threads not tested because creating many instances of those triggers other intermediary objects to be sampled so couldn’t easily get to the `Thread` instances. Classes not tested since it requires dynamically creating brand new `Class` instances at runtime, something not feasible in closed-world SubstrateVM.

Once this PR is integrated I'll work on adding path to GC root functionality.

galderz commented:
Can @roberttoyonaga @christianhaeubl and @adinn review this?

galderz commented:
Fixing some eclipse format issues...

galderz commented:
Added a few commits to address warning errors. Also made the full sample queue test more reliable.

galderz commented:
I think I've fixed all style and format errors. However, gate testing is not working. It fails with:

```
Tag "test" not part of any task.
Run the following command to see all available tasks and their tags:
  mx -v gate --dry-run
Traceback (most recent call last):
  File "/home/runner/work/graal/graal/mx/mx_gate.py", line 475, in gate
    mx.abort(f'Tag "{tag}" not part of any task.\n'
  File "/home/runner/work/graal/graal/mx/mx.py", line 4273, in abort
    raise SystemExit(error_code)
SystemExit: 1
```

@christianhaeubl Any ideas? @zakkak said it could be related to https://github.com/oracle/graal/pull/6738.

zakkak commented:
> @zakkak said it could be related to #6738.

FTR I have created https://github.com/oracle/graal/pull/6746 to resolve the issue.

roberttoyonaga reviewed (commented):
Thanks Galder! It would be really awesome to add this support for JFR leak profiling. I've left a few comments/questions here.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/locks/SpinLock.java:
```diff
@@ -0,0 +1,100 @@
+    public void lock() {
+        @SuppressWarnings("deprecation")
+        final long threadId = Thread.currentThread().getId();
+        while (!lock.compareAndSet(NOT_HELD, threadId)) {
```
Good catch. I had completely missed that class. I've just tried it and does the job.

galderz commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/oldobject/JfrOldObjectTest.java:
```diff
@@ -0,0 +1,118 @@
+    public void triggerUpdateOfLastKnownHeapUsage() {
+        // Trigger GC before tests to get a reading of last known heap usage for the first executed
+        // test.
+        System.gc();
```
Yup, once is enough :)

galderz commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/oldobject/JfrOldObjectTest.java:
```diff
@@ -0,0 +1,118 @@
+        return startRecording(events, getDefaultConfiguration(), settings);
+    }
+
+    static void blackhole(Object obj) {
```
Good question. The `leak` field is assigned to a static variable but it's not used, so at the beginning when I started working on this I was worried the compiler might optimize this away and remove all the code. The `blackhole` method is just a way to make sure the field is used in a trivial way, but one that cannot be optimized away because it can (extremely unlikely) end up causing a side effect (printing an empty space character).

I've just removed it and calls to it and it all works fine, so I'll go ahead and remove it.

galderz commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/oldobject/TestArrayLeak.java:
```diff
@@ -0,0 +1,52 @@
+            Object[] right = new Object[3];
+            node[2] = right;
+            node = right;
+        }
```
> Is this adding objects to this array try to make the 100 element array allocation get sampled along the allocation slow path? If so, is there a chance that it doesn't actually get sampled?

All I'm doing is generating enough arrays so that some of them get sampled as old objects and I can verify that sampled arrays make it through as events and they have the right data populated. Having the right data here means that the sampled event is for Object[] has length of 100. If no arrays were sampled the test would fail because it guarantees that the list of events generated for the `Object[]` is not empty.

> What if you directly called JfrOldObjectSampleEvents.sample(result, size.rawValue(), Integer.MIN_VALUE); from this testing code?

Hmmm, not sure what that would do that this test is not already covering?

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrChunkWriter.java:
```diff
@@ -106,7 +107,7 @@ public JfrChunkWriter(JfrGlobalMemory globalMemory, JfrStackTraceRepository stac
          * changing epoch.
          */
-        this.flushCheckpointRepos = new JfrRepository[]{stackTraceRepo, methodRepo, typeRepo, symbolRepo};
+        this.flushCheckpointRepos = new JfrRepository[]{stackTraceRepo, methodRepo, typeRepo, symbolRepo, oldObjectRepo};
```
Makes sense, I've reordered repos.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectSampler.java:
```diff
@@ -0,0 +1,172 @@
+        int numDead = 0;
+        for (int i = 0; i < samples.getCapacity(); i++) {
+            final OldObject sample = samples.getSample(i);
+            if (sample.reference != null && sample.reference.get() == null) {
```
Yeah, small changes indeed :)

I've not pushed all the commits yet because I was trying to get clarification about https://github.com/oracle/graal/pull/6743#discussion_r1223552890, but that might take some days, so I'll probably push updates shortly.

galderz commented:
Pushed a few updates that should address all of @roberttoyonaga's feedback:

* Use `JavaSpinLockUtils` instead of rolling your own.
* Make calls to `WeakReference.get()` uninterruptible. To achieve this, since `Reference` class is substituted,
annotation processor had to be changed to enable `@KeepOriginal` and `@AnnotateOriginal` to be used together
in classes that are marked with `@Substitute`.
* Just called `System.gc()` call once to record a first heap usage.
* Remove `blackhole` methods which are not necessary.
* Put old object repo before type and symbol repo.
* Support flush of old object repository.
* Link samples with a addition-time-based singly linked list, in order to keep same logic as in HotSpot when scavenging. When removing entries through scavenging, their priority is passed on to an sample allocated just after it. This is done to keep samples distributed evenly.
* Migrate sampler to profiler and separate sampler and emitter. By doing that, the profiler can be more easily unit tested, while still encapsulating the sampler, emitter, queue and list implementations.
* Added a unit test for the old object profiler.
* Just like in HotSpot, add total allocated tracking logic.
* Made clear which of the tests exercise JFR recording vs the other unit tests.

galderz commented:
Formatting, style...etc should be done now, but CI shows all the recording tests I've added failing when tested with `labsjdk-ce-21-jvmci-23.1-b09`. The problem arises because `object` field of the recorded event seems to be null. E.g.

```
jdk.OldObjectSample {
  startTime = 13:40:06.910 (2023-06-27)
  allocationTime = 13:40:06.910 (2023-06-27)
  objectSize = 33.8 kB
  objectAge = 19.2 ms
  lastKnownHeapUsage = 216 bytes
  object = N/A
...
```

I'm looking into this.

galderz commented:
Hmmmm, seems like in JDK 21 (or some other version post 17) the event has an extra field `objectSize`:

```
            E: field
             A: name allocationTime
             A: class 219
                E: annotation
                 A: class 18707
                E: annotation
                 A: class 18699
                 A: value TICKS
                E: annotation
                 A: class 7632
                 A: value Allocation Time
            E: field
             A: name objectSize
             A: class 219
                E: annotation
                 A: class 18707
                E: annotation
                 A: class 18720
                 A: value BYTES
                E: annotation
                 A: class 7632
                 A: value Object Size
            E: field
             A: name objectAge
             A: class 219
                E: annotation
                 A: class 18707
                E: annotation
                 A: class 18693
                 A: value TICKS
                E: annotation
                 A: class 7632
                 A: value Object Age
```

Compared with JDK17:

```
            E: field
             A: name allocationTime
             A: class 205
                E: annotation
                 A: class 18875
                E: annotation
                 A: class 18866
                 A: value TICKS
                E: annotation
                 A: class 6859
                 A: value Allocation Time
            E: field
             A: name objectAge
             A: class 205
                E: annotation
                 A: class 18875
                E: annotation
                 A: class 18860
                 A: value TICKS
                E: annotation
                 A: class 6859
                 A: value Object Age
```

Observed parsing the JFR file with @jiekang's excellent [jfr-reader tool](https://github.com/jiekang/jfr-reader).

I'm unsure how we should handle this. I'll see if I can handle it using the JDK earlier/later boolean suppliers.

galderz commented:
FYI `objectSize` field was [added](https://bugs.openjdk.org/browse/JDK-8269092) in JDK18.

galderz commented:
Fixed tests and pushed commits to support JDK18+. I've ended up removing object description tests because in later JDKs generating `ThreadGroup` instances unfortunately creates weak reference arrays that hide `ThreadGroup` instance samples (e.g. see [here](https://gist.github.com/galderz/fca2238bb1f2bc763aef35ff7e64f4ed)). This changes have been introduced via the virtual thread support. I had similar issues with other types created while creating `Thread` instances. As well as being pretty heavyweight objects, creating these classes generate a lot of other garbage which dominates samples. I will defer testing of object descriptions, via unit tests, to the path to GC root work, because that involves considerable changes on the emit side, so I can use that opportunity to revisit the code more thoroughly and find an easy way to validate these descriptions.

roberttoyonaga reviewed (commented):
Thanks for the updates @galderz! I think you've addressed all my earlier comments. I left a few more pretty minor notes.

roberttoyonaga commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/heap/Target_java_lang_ref_PhantomReference.java:
```diff
@@ -36,4 +38,8 @@ public final class Target_java_lang_ref_PhantomReference<T> {
     }
+
+    @AnnotateOriginal
+    @Uninterruptible(reason = "Called from uninterruptible code.", mayBeInlined = true)
```
I understand why you need it for `Reference`, but why do you need to add the uninterruptible annotation to this class?

galderz replied:
Good question. I don't know the exact details on the limitations around `@Uninterruptible`, but unless you do this native image fails, e.g.:

```
com.oracle.svm.core.util.VMError$HostedError: Found 1 violations of @Uninterruptible usage:
- method java.lang.ref.Reference.get():Object is annotated but java.lang.ref.PhantomReference.get():Object is not
```

I suspect that if classes override a method that is `@Uninterruptible`, e.g. `PhantomReference`, then they also need to mark it `@Uninterruptible`.

galderz commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/oldobject/TestOldObjectProfiler.java:
```diff
@@ -0,0 +1,384 @@
+            profiler.sample(new WeakReference<>(String.valueOf(i)), i * 100, -1);
+        }
+
+        profiler.emit(0, Long.MAX_VALUE);
```
Hmmmm, I think I know what you mean, but let's refresh how scavenging has been ported to see if this makes sense:

Technically the way `scavenge` works in HotSpot is like this:

```
void ObjectSampler::add(HeapWord* obj, size_t allocated, traceid thread_id, bool virtual_thread, const JfrBlobHandle& bh, JavaThread* thread) {
  ...

  if (Atomic::load(&_dead_samples)) {
    // There's a small race where a GC scan might reset this to true, potentially
    // causing a back-to-back scavenge.
    Atomic::store(&_dead_samples, false);
    scavenge();
  }
```

It only gets called when sampled, but it calls it whenever a sample has been GC'd. There's no such notification mechanism in SVM, so instead I went for the option of scavenging when the queue is full.

Now, the issue I think you're implying above is that if the queue is not full, no scavenging happens and maybe it's good idea to scavenge after emitting to see if anything is dead and can be removed. This sounds like a good idea to me. It seems better than an alternative option of scavenging in each object sample add call. Have I got this right?

galderz replied:
> You may not even need to run a full `scavenge()` operation. When emitting in `OldObjectEventEmitter.emitUnchained(long)` you already check if each sample is alive. If it is dead, you can directly remove it with `OldObjectSampler.remove(OldObject)`. Admittedly, this is only really useful if you have multiple simultaneous recordings running since we only emit the OldObjectSample event once per recording anyway (so usually there will be no point to doing the extra cleanup work during event emission).

Ok, I think I'll leave the logic as is for now.

> hmm my original comment was actually concerning the test itself. I think that your current implementation of scavenging/emitting works perfectly fine. I think `testScavengeMiddle`, `testScavengeYoungest`, `testScavengeOldest` and `testScavengeAll` don't actually exercise the code in `OldObjectSampler.scavenge()` and instead rely on `OldObjectEventEmitter.emitUnchained(long)` to remove the dead samples.

That's right. I'll rework this to make sure the scavenging gets tested.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectSampler.java:
```diff
@@ -0,0 +1,140 @@
+            // the weight of a sample that is removed should be redistributed.
+            // Here it gets redistributed to the sample that came just after in time.
+            queue.remove(prev);
+            prev.span += sample.span;
```
Another interesting observation which also needs some further explanation :)

Because the span is added it to the younger neighbour, that is why I implemented it as a single-linked list. In HotSpot, this is implemented as double-linked list so it could efficiently distribute the span to the older sample. I went for the simpler, more space efficient, option, taking into account that the default sample size is relatively small, so when removing things I can just loop around and fix things up.

Now, about how to distribute the span. [Here's what Erik said recently when I asked him about the linked list of samples](https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005067.html):

> The weight of the object, the allocation span, in the time-based list, should be divided to its neighbour(s).

Then [@tstuefe pointed out](https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005069.html):

> span is redistributed to its (in time) predecessor sample. Following your explanation, should the span not be evenly divided between predecessor and successor?

The first part is not entirely correct since `prev` is really the sample that was sampled after the current object, so indeed the younger one. But the point is that Erik suggest it should be divided. So, 50% to younger and 50% to older. Unfortunately Erik didn't reply to this, so I think it would make sense to follow up on hotspot-jfr-dev and see if HotSpot should change accordingly. Feel free to kick off a discussion there since I'll be away for a bit.

roberttoyonaga replied:
Thanks for the detailed response! Ok that all makes sense to me. I agree that a singly-linked is the best option to achieve the same behavior as in hotspot (only redistribute to the younger neighbor).

That's a good idea. I'll follow up on your "hotspot-jfr-dev" email thread asking "why doesn't the Hotspot implementation actually distribute the span evenly to both neighbors? And if it must distribute unevenly, why not choose the older neighbor?".

roberttoyonaga replied:
I asked on the hotspot mailing list here: https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005164.html

Unfortunately, there was no response.

roberttoyonaga reviewed (approved):
Thanks for the new changes. It looks good to me.

galderz commented:
The espresso and sulong failures seem unrelated.

galderz commented:
@christianhaeubl Any thoughts on this PR?

galderz commented:
FYI I've just updated the PR to fix the conflicts.

galderz commented:
Updated the code to fix conflicts. As a result of changes in master, I moved the code to using `ReferenceInternals.getReferent` to get the value out of the weak reference. The method I had used before to add uninterruptible to `Reference.get` methods didn't work any more.

By the way, let me know when you're happy with the code and I can squash all these commits. This would greatly simplify potential backports down the line.

christianhaeubl commented on substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/substitute/AnnotationSubstitutionProcessor.java:
```diff
@@ -772,7 +772,13 @@ private void handleAnnotatedMethodInSubstitutionClass(Executable annotatedMethod
             register(methodSubstitutions, annotated, original, substitution);
         } else if (keepOriginalAnnotation != null) {
-            register(methodSubstitutions, annotated, original, original);
+            AnnotateOriginal annotateOriginalAnnotation = lookupAnnotation(annotatedMethod, AnnotateOriginal.class);
```
Do you still need this change?

galderz replied:
Indeed not needed.

christianhaeubl commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectSampler.java:
```diff
@@ -0,0 +1,139 @@
+        if (queue.isFull()) {
+            if (queue.peek().span > span) {
+                // Sample will not fit, try to scavenge
+                // HotSpot code scavenges when it gets notified that a sample was GC'd,
```
Theoretically, you could use a reference queue to get notified but I think that is too tricky to do from uninterruptible code. How does the notification work in HotSpot?

christianhaeubl commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectProfiler.java:
```diff
@@ -0,0 +1,144 @@
+
+    // Moving locking to OldObjectProfiler class cannot easily be done.
+    // For starters Unsafe needs to be accessible there,
+    // but there might be other issues.
```
For multi-line comments, we are moving more and more towards `/* */`-style comments to avoid formatting issues.

christianhaeubl commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectSampler.java:
```diff
@@ -0,0 +1,139 @@
+final class OldObjectSampler {
+    private final OldObjectArray samples;
+    private final OldObjectPriorityQueue queue;
+    private final OldObjectList list;
```
Please add some class-level JavaDoc why the 3 different views on the data are necessary (array, list, priority queue).

christianhaeubl reviewed (commented):
Thanks for the PR, I reviewed most of it (except for test cases) and added questions and comments.

galderz commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GCImpl.java:
```diff
@@ -208,17 +210,21 @@ private void collectOperation(CollectionVMOperationData data) {
+        Heap.getHeap().updateUsedAtGC();
         JfrGCHeapSummaryEvent.emit(JfrGCWhen.AFTER_GC);
 
+        UnsignedWord sizeAfter = heapAccounting.getEdenUsedBytes().add(heapAccounting.getSurvivorUsedBytes()).add(heapAccounting.getOldUsedBytes());
```
Ah yes. Initially I didn't spot that survivor bytes are counted `youngUsedBytes`, but I see now that after in after collection eden and survivor are added together and they update `youngUsedBytes`.

galderz commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/GCImpl.java:
```diff
@@ -208,17 +210,21 @@ private void collectOperation(CollectionVMOperationData data) {
         GenScavengeMemoryPoolMXBeans.notifyAfterCollection();
 
         printGCAfter(cause);
+        Heap.getHeap().updateUsedAtGC();
```
The problem with this approach is in `HeapAccounting`, which is not accessible outside its module. I need to get the heap used after last GC from the svm.core module. This is why I added the methods directly to `Heap`. Any ideas around this? I doubt you want to open up `HeapAccounting` to access externally? `getUsedAtLastGC`, or a similarly named method, could remain in `Heap` and then define `updateUsedAtGC` in `HeapImpl`.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectProfiler.java:
```diff
@@ -0,0 +1,144 @@
+
+    private int queueSize;
+    private long lastSweep = Long.MAX_VALUE;
+    private OldObjectProfiler profiler;
```
This is where looking at the tests helps 😄 

The reason I separated these two classes is to enable me to quickly test the profiler logic in JVM mode. I did this for quicker iteration developing and debugging the algorithms in the profiler. Ideally I would like to keep this separation. 

`TestOldObjectProfiler` is the class that tests `OldObjectProfiler` in isolation and it's runnable in both JVM and native mode. I did attempt to merge `JfrOldObjectProfiler` and `OldObjectProfiler` together, but I couldn't get the locking part to work when running the class in pure JVM mode.

Thoughts?

galderz replied:
Ended up removing `OldObjectProfiler`.

galderz commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrOldObjectSampleEvents.java:
```diff
@@ -0,0 +1,62 @@
+
+import java.lang.ref.WeakReference;
+
+public class JfrOldObjectSampleEvents {
```
> The class name confused me a bit. I don't think that this class should have Events in its name because it is only used for the sampling (and not for emitting the event).

Makes sense. I can rename it to `JfrOldObjectSampler`.

> allocatedSize should be of type UnsignedWord (if a conversion from UnsignedWord to long is necessary, we try to do it as late as possible).

Hmmmm, this clashes a bit with my attempts to test the profiler logic in a way that it can run in JVM mode (see https://github.com/oracle/graal/pull/6743#discussion_r1335373865). Is it possible to run code that uses `UnsignedWord` in JVM mode? By doing an early switch to long I am then able to keep the `OldObjectProfiler` logic runnable in both JVM and native modes.

galderz commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrOldObjectSampleEvents.java:
```diff
@@ -0,0 +1,62 @@
+        if (HasJfrSupport.get()) {
+            // Instantiate weak reference at the last possible time before allocations are not
+            // allowed
+            sample(new WeakReference<>(obj), allocatedSize, arrayLength);
```
Hmmm, I see this pattern has now morphed into `needsVMOperation()`. I assume it is still applicable? I'll see how it can be applied.

galderz replied:
I got this working.

galderz commented on substratevm/src/com.oracle.svm.core.genscavenge/src/com/oracle/svm/core/genscavenge/JfrOldObjectSampleEvents.java:
```diff
@@ -0,0 +1,62 @@
+    static void updateLastSweep(UnsignedWord sizeBefore, UnsignedWord sizeAfter) {
+        final long timestamp = JfrTicks.elapsedTicks();
+        if (timestamp > 0 && HasJfrSupport.get()) {
+            if (sizeAfter.belowThan(sizeBefore)) {
```
This tries to replicate the HotSpot logic in `objectSampler.cpp`:

```
// Callback invoked by the GC after an iteration over the oop storage
// that may have cleared dead referents. num_dead is the number of entries
// already nullptr or cleared by the iteration.
void ObjectSampler::oop_storage_gc_notification(size_t num_dead) {
  if (num_dead != 0) {
    // The ObjectSampler instance may have already been cleaned or a new
    // instance was created concurrently.  This allows for a small race where cleaning
    // could be done again.
    Atomic::store(&_dead_samples, true);
    Atomic::store(&_last_sweep, (int64_t)JfrTicks::now().value());
  }
}
```

The last sweep time is only updated if any objects were GC'd. I couldn't see a way to do exactly that on svm, so I used sizes to decide whether GC freed up any memory or not. I don't know how likely is for GC to occur and nothing to be freed.

The aim of tracking last sweep time is to enable the old object sampler feature  where only samples that are older than last GC are emitted. This happens when emit is called with `emitAll` being `false`. When is `true` then all samples are emitted and no distinction is made.

galderz replied:
This optimization will be removed for now, until #7503 is in place.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrNativeEventWriter.java:
```diff
@@ -223,6 +223,51 @@ public static void putString(JfrNativeEventWriterData data, String string, CharR
     }
 
+    @Uninterruptible(reason = "Accesses a native JFR buffer.", callerMustBe = true)
+    public static void putString(JfrNativeEventWriterData data, String string, int stringLengthLimit, String prefix) {
```
I went for the option of using a pre-allocated `char[]`. I was then able to remove this `putString`, but I added a `putChars` that took the `char[]` buffer and a length of how much of it is written. It also required added a `modifiedUTF8Length(char[] chars, int length)` and `toModifiedUTF8(int lenght, char[] chars...etc)`.

I was also able to unit test this without pollution by directly hitting the `sample` method of the old object sampler. The coverage is for thread group, thread and class and includes an ellipsis description test.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/JfrNativeEventWriter.java:
```diff
@@ -223,6 +223,51 @@ public static void putString(JfrNativeEventWriterData data, String string, CharR
+            putByte(data, JfrChunkWriter.StringEncoding.EMPTY_STRING.getValue());
+        } else {
+            Pointer newPosition;
+            if (string != null && string.length() > stringLengthLimit - 3) {
```
No longer applicable due to changes related to https://github.com/oracle/graal/pull/6743#discussion_r1350133166.

galderz commented:
I pushed commits too early, I still have some comments to handle. I'll post an update when they're ready for another review.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectProfiler.java:
```diff
@@ -0,0 +1,144 @@
+    @SuppressWarnings("unused") private volatile int lock;
+
+    private int queueSize;
+    private long lastSweep = Long.MAX_VALUE;
```
This instance variable and related logic has been removed for now as per https://github.com/oracle/graal/pull/6743#discussion_r1335720928.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectRepository.java:
```diff
@@ -0,0 +1,172 @@
+            final Word pointer = Word.objectToUntrackedPointer(obj);
+            JfrNativeEventWriterData data = StackValue.get(JfrNativeEventWriterData.class);
+            JfrNativeEventWriterDataAccess.initialize(data, epochData.buffer);
+            final long objectId = nextId++;
```
The initial value is 0 and it's increment before using it, so the first value used would be 1. It would be clearer though if the value was initialized to 1 and incremented after use. I'll amend that.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/JfrOldObjectRepository.java:
```diff
@@ -0,0 +1,172 @@
+            return;
+        }
+
+        // Size description not implemented since that relies on runtime reflection.
```
Created follow-up issue https://github.com/oracle/graal/issues/7569.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectEventEmitter.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Uninterruptible(reason = "Prevent JFR recording and epoch change.")
+    private static boolean isOlderThan(long lastSweep, long allocationTime) {
```
No longer relevant after removing partial emit logic https://github.com/oracle/graal/pull/6743#discussion_r1335720928.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectEventEmitter.java:
```diff
@@ -0,0 +1,77 @@
+
+import com.oracle.svm.core.Uninterruptible;
+
+final class OldObjectEventEmitter {
```
Right it doesn't have much but it will expand once path-to-gc-root functionality is added, which would go here. I also liked the idea of separating it so that it's clear that for emitting only the list of view of the old object samples is needed. Then becomes is clear that the queue view it's only used for sampling.

galderz commented:
@christianhaeubl I've addressed all comments. Can you have another look? As said above, let me know when you're happy before integrating it so that I can squash all commits. Doing this makes it easier for me to backport it to another branch if necessary.

galderz commented:
I've fixed the conflicts due to graal compiler package refactoring and I used the opportunity to squash the commits.

galderz commented:
Fixed another conflict and pushed.

christianhaeubl reviewed (commented):
@galderz : I am currently cleaning up this PR so that we can integrated into master. As part of that, I came across a few code parts that were unclear, so I added a few questions.

roberttoyonaga commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectSampler.java:
```diff
@@ -0,0 +1,147 @@
+    @Uninterruptible(reason = "Access protected by lock.")
+    public boolean sample(WeakReference<Object> ref, long allocatedSize, int arrayLength) {
+        totalAllocated += allocatedSize;
+        long span = totalAllocated - queue.total();
```
This is also what Thomas Stuefe was concerned about. See mailing list message: https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005069.html

As far as I can tell, this question wasn't given a clear answer.

galderz commented on substratevm/src/com.oracle.svm.core/src/com/oracle/svm/core/jfr/oldobject/OldObjectEventEmitter.java:
```diff
@@ -0,0 +1,76 @@
+        OldObject current = list.head();
+        while (current != null) {
+            if (current.reference != null) {
+                final Object obj = effects.getWeakReferent(current.reference);
```
> Two questions:
> 
> * After emitting the data, do we need to clear the data structures?

Which data structure are you talking about exactly? The sample queue/list has to remain in place, the elements in it only get cleared when they're GC'd. So, they keep being emitted until then.

> * Currently, we emit the data in the order from the oldest to the youngest allocation. Is it necessary to emit the data in this order?

I've been wondering about this and I explicitly asked about this in the jfr-dev list (see [discussion](https://mail.openjdk.org/pipermail/hotspot-jfr-dev/2023-June/005066.html)).

Specifically I asked:

> I'm wondering now if the fact that the list view is ordered by insertion time is relevant here. Is it maybe trying to give more importance (by increasing its span) to surviving objects that are near in time to those that have been garbage collected.

Erik replied:

> The intent of the code is to remove the sample from both the time-based list and the priority queue as the object is no longer alive.

So the time aspect of the list order is acknowledged.

Now, when it comes to emitting the events, the HotSpot code uses the list order to write the events. See [here](https://github.com/openjdk/jdk/blob/master/src/hotspot/share/jfr/leakprofiler/checkpoint/eventEmitter.cpp#L97-L105). `object_sampler->first()` calls into `_list->first()`. I've tried to replicate the same logic in svm.

galderz commented:
> What was the final outcome of this discussion?
> What if all sampled objects are dead at some point? Then, we don't redistribute the span and queue.total() will in fall to 0. This will result in a large difference between totalAllocated and queue.total(), so new samples can easily evict old samples for a long period of time.

There has not been further replies to this discussion. I've replied in the thread to see if Erik can provide an answer.

In any case, if it's really a bug, then we wait for the hotspot fix and then backport it here. That way the two impls are relatively close in their logic.

galderz commented:
@christianhaeubl I see there's a conflict in `HeapImpl`. Shall I address it or are you taking care of it?

christianhaeubl commented:
@galderz: I will handle the conflict.

christianhaeubl commented on substratevm/src/com.oracle.svm.test/src/com/oracle/svm/test/jfr/oldobject/TestRecordingArrayLeak.java:
```diff
@@ -0,0 +1,62 @@
+        }
+        // Trigger a GC so that last sweep gets updated,
+        // and the objects above are considered older than last GC sweep time.
+        System.gc();
```
One, probably last question: what is the intention of this line? I don't think that your code accesses the last GC sweep time anywhere.

galderz replied:
Good catch. I think this is leftover from previous versions of the code when I was tracking the last GC sweep time, but that is only relevant when path to GC root is on, and you set a cut off time for the samples. I'll remove it and make sure everything works as expected.

christianhaeubl commented:
@galderz , @roberttoyonaga : for integration into master, I opened https://github.com/oracle/graal/pull/8251 which adds a larger number of cleanups and fixes smaller bugs. Feel free to review this PR.

galderz commented:
@christianhaeubl Now that https://github.com/oracle/graal/pull/8251 is integrated, shall we close this?

christianhaeubl commented:
Yes, good point.
